dreadnode/agent/tools/fs.py

```diff
@@ -253,39 +284,85 @@ def grep(
                         for j in range(context_start, context_end):
                             prefix = ">" if j == i else " "
                             line_text = lines[j].rstrip("\r\n")
-                            context.append(f"{prefix} {j + 1}: {shorten_string(line_text, 80)}")
+                            context.append(
+                                f"{prefix} {j + 1}: {shorten_string(line_text, 80)}"
+                            )
 
                         rel_path = self._relative(file_path)
-                        matches.append(
+                        file_matches.append(
                             GrepMatch(
                                 path=rel_path,
                                 line_number=line_num,
                                 line=shorten_string(line.rstrip("\r\n"), 80),
                                 context=context,
                             ),
                         )
+            except (
+                FileNotFoundError,
+                PermissionError,
+                IsADirectoryError,
+                UnicodeDecodeError,
+                OSError,
+            ) as e:
+                logger.warning(f"Error occurred while searching file {file_path}: {e}")
+
+            return file_matches
+
+        # Search files in parallel with concurrency limit
+        semaphore = asyncio.Semaphore(self.max_concurrent_reads)
+
+        async def search_file_limited(file_path: UPath) -> list[GrepMatch]:
+            """Search a single file with semaphore to limit concurrency."""
+            async with semaphore:
+                return await search_file(file_path)
+
+        all_matches: list[GrepMatch] = []
+        results = await asyncio.gather(
+            *[search_file_limited(file_path) for file_path in files_to_search]
+        )
+
+        # Flatten results and respect max_results
+        for file_matches in results:
+            all_matches.extend(file_matches)
+            if len(all_matches) >= max_results:
+                break
 
-        return matches
+        return all_matches[:max_results]
 
     @tool_method(variants=["write"], catch=True)
-    def write_file(
+    async def write_file(
         self,
         path: t.Annotated[str, "Path to write the file to"],
         contents: t.Annotated[str, "Content to write to the file"],
     ) -> FilesystemItem:
         """Create or overwrite a file with the given contents."""
-        _path = self._safe_create_file(path)
-        with _path.open("w") as f:
-            f.write(contents)
+        _path = await self._safe_create_file(path)
+        async with aiofiles.open(_path, "w") as f:
+            await f.write(contents)
```

Comment:
Just double checking here that aiofiles supports UPath correctly - we want to be able to pass things like `s3://` still.